numpy/lib/format.py

```diff
@@ -242,6 +242,19 @@ def read_magic(fp):
         major, minor = magic_str[-2:]
     return major, minor
 
+class MetadataWarning(UserWarning):
+    pass
+
+def _has_metadata(dt):
+    if dt.metadata is not None:
+        return True
+    elif dt.names is not None:
+        return any(_has_metadata(dt[k]) for k in dt.names)
+    elif dt.subdtype is not None:
+        return _has_metadata(dt.base)
```

Comment:
This is a subarray dtype, such as `np.dtype((int, 3), metadata={})`

Reply:
Where is the metadata in `xxx = np.dtype((int, 3), metadata={'some': 'stuff'})` stored? It is neither in `xxx.metadata` nor in `xxx.subdtype[0].metadata`

Reply:
The meta data is not preserved in creating an array with that dtype:
```
>>> dt=np.dtype((int, 3), metadata={'some': 'stuff'})
>>> dt.metadata
mappingproxy({'some': 'stuff'})>>> arr = np.ones(10, dtype=dt)
>>> arr.dtype.metadata
None
```

Reply:
Subarray types have a short half-life unless protected by a containing structured type :)